src/Microsoft.ML.Transforms/Text/CharTokenizeTransform.cs

```diff
@@ -405,34 +405,55 @@ private ValueGetter<VBuffer<ushort>> MakeGetterVec(IRow input, int iinfo)
                     getSrc(ref src);
 
                     int len = 0;
+                    
                     for (int i = 0; i < src.Count; i++)
                     {
                         if (src.Values[i].HasChars)
                         {
                             len += src.Values[i].Length;
-                            if (_useMarkerChars)
-                                len += TextMarkersCount;
+
+                            if (i > 0)
+                                len += 1;  // add space character that will be added
                         }
                     }
 
+                    if (_useMarkerChars)
+                        len += TextMarkersCount;
+
                     var values = dst.Values;
                     if (len > 0)
                     {
                         if (Utils.Size(values) < len)
                             values = new ushort[len];
 
                         int index = 0;
+
+                        // VBuffer<DvText> can be a result of either concatenating text columns together 
+                        // or application of word tokenizer before char tokenizer.
+                        //
+                        // Considering VBuffer<DvText> as a single text stream.
+                        // Therefore, prepend and append start and end markers only once i.e. at the start and at end of vector.
+                        // Insert spaces after every piece of text in the vector.
+                        if (_useMarkerChars)
+                            values[index++] = TextStartMarker;
+
                         for (int i = 0; i < src.Count; i++)
                         {
                             if (!src.Values[i].HasChars)
                                 continue;
-                            if (_useMarkerChars)
-                                values[index++] = TextStartMarker;
+
+                            if (i > 0)
+                                values[index++] = ' ';
```

Comment:
Right. Space was *not* selected as the breaking character between tokens in a vector explicitly because it was quite likely to occur within individual tokens. This is why we chose a character unlikely to occur in text (vs. ***one of the most frequently observed characters***) in the first place. :)

---
In reply to: [203213571](https://github.com/dotnet/machinelearning/pull/548#discussion_r203213571) [](ancestors = 203213571)

Reply:
So, there is no way to get the token that was used for tokenization?

Reply:
In reality, if an statistical technique for word tokenization is applied then there is no reliable way to get the actual text after it has been tokenized. 

So, use of space is just an approximation. Otherwise, it not possible to get actual text string.

Reply:
> So, there is no way to get the token that was used for tokenization?

No. Indeed you have absolutely no idea whether the tokenization scheme involved breaking on single characters at all -- a vector of text could have come from anywhere. Indeed, more elaborate tokenization schemes exist, that involve an actual language model and so are considerably more elaborate than a simple `string.Split(char)`, which of course is the assumption made by this "fix."

I suggest that we not try to get into the business of trying to guess *how* the data that was fed into a transform, since doing so degrades the system. It would do more harm than good to do it -- we've built a modular system of transforms and data pipelines, but the consequence of a modular transform is that you must accept that your data *could* come from literally anywhere.

What I believe is a reasonable compromise, is that some sort of special character... perhaps a character not used often in text, but that is nonetheless specially reserved to indicate the start and end of a given block of text.

Happily, I believe that was the state of affairs of this transform in its existing implementation, before it was "fixed." :smile: